File: domain/src/main/java/com/arturogutierrez/openticator/domain/category/interactor/AddAccountToCategoryInteractor.java
package com.arturogutierrez.openticator.domain.category.interactor;

import com.arturogutierrez.openticator.domain.account.model.Account;
import com.arturogutierrez.openticator.domain.category.model.Category;
import com.arturogutierrez.openticator.domain.category.repository.CategoryRepository;
import com.arturogutierrez.openticator.executor.PostExecutionThread;
import com.arturogutierrez.openticator.executor.ThreadExecutor;
import com.arturogutierrez.openticator.interactor.Interactor;
import rx.Observable;

public class AddAccountToCategoryInteractor extends Interactor<Category> {

  private final CategoryRepository categoryRepository;

  private Category category;
  private Account accountToAddToCategory;

  public AddAccountToCategoryInteractor(CategoryRepository categoryRepository,
      ThreadExecutor threadExecutor, PostExecutionThread postExecutionThread) {
    super(threadExecutor, postExecutionThread);
    this.categoryRepository = categoryRepository;
  }

  public void configure(Category category, Account account) {
    this.category = category;
    this.accountToAddToCategory = account;
  }

  @Override
  public Observable<Category> createObservable() {
    if (accountToAddToCategory == null) {
      throw new IllegalStateException("You must call configure before execute the interactor");
    }

    return categoryRepository.addAccount(category, accountToAddToCategory);
  }
}

File: domain/src/main/kotlin/com/arturogutierrez/openticator/domain/category/interactor/AddAccountToCategoryInteractor.kt
package com.arturogutierrez.openticator.domain.category.interactor

import com.arturogutierrez.openticator.domain.account.model.Account
import com.arturogutierrez.openticator.domain.category.interactor.AddAccountToCategoryInteractor.Params
import com.arturogutierrez.openticator.domain.category.model.Category
import com.arturogutierrez.openticator.domain.category.repository.CategoryRepository
import com.arturogutierrez.openticator.executor.PostExecutionThread
import com.arturogutierrez.openticator.executor.ThreadExecutor
import com.arturogutierrez.openticator.interactor.SingleUseCase
import io.reactivex.Single

class AddAccountToCategoryInteractor(private val categoryRepository: CategoryRepository,
                                     threadExecutor: ThreadExecutor,
                                     postExecutionThread: PostExecutionThread)
  : SingleUseCase<Category, Params>(threadExecutor, postExecutionThread) {

  override fun buildObservable(params: Params): Single<Category> {
    val category = params.category
    val accountToAddToCategory = params.account
    return categoryRepository.addAccount(category, accountToAddToCategory)
  }

  data class Params(val category: Category, val account: Account)
}
